### PR TITLE
[Nuclio] Add NS enrichment to Nuclio config

### PR DESCRIPTION
**Pull request by yaelgen**
This PR introduces node selector enrichment for Nuclio config within MLRun. MLRun will now incorporate node selector enrichments directly into the Nuclio config, merging settings from both the project level and the MLRun service level. 
In case of any overlap, node selectors from the MLRun service level will take precedence over those specified at the Nuclio service level.

Although this change is not tied to a specific ticket, it aligns with the approach outlined in [this ticket](https://iguazio.atlassian.net/browse/ML-7290). The agreed approach ensures that MLRun effectively manages node selector enrichment when deploying Nuclio functions through MLRun.

### CODE REVIEW

File: server/api/crud/runtimes/nuclio/function.py [hunk 1]
```diff
@@ -493,6 +493,24 @@ def _resolve_and_set_build_requirements_and_commands(function, config):
...
+    if run_db and project_name:
+        if project := run_db.get_project(project_name):
+            project_node_selector = project.spec.default_function_node_selector
+
+    # If either project node selector or config node selector is defined, enrich the config node selector
+    if project_node_selector or config_node_selector:
+        return mlrun.utils.helpers.merge_dicts_with_precedence(
+            config_node_selector, project_node_selector, function_node_selector
+        )
+    return function_node_selector
```
Comment (TomerShor) [inline server/api/crud/runtimes/nuclio/function.py]: Is there a reason we can't use one of the multiple other versions of this function that already exist?

I know some of the implementations have some runtime-specific intricacies, but adding the same functionality in each runtime is not maintainable at all.
See how can you common-ize these functions as much as you can.

File: tests/api/runtimes/test_nuclio.py [hunk 2]
```diff
@@ -923,6 +923,34 @@ def test_deploy_with_node_selection(self, db: Session, client: TestClient):
             tolerations=tolerations,
         )
 
+    def test_compile_function_config_node_selector_enriched_from_project(
```
Comment (TomerShor) [inline tests/api/runtimes/test_nuclio.py:936]: do you want to test subsets (only project / only config) or is it already tested in UTs for `merge_dicts_with_precedence` ?

**Review by TomerShor — CHANGES_REQUESTED** [review]

File: tests/api/runtimes/test_nuclio.py (hunk 2 above)
Comment (yaelgen) [reply tests/api/runtimes/test_nuclio.py:936]: In my opinion it is already covered with the UTs for `merge_dicts_with_precedence`, but maybe I will add a case for not having any NS defined to make sure nothing is broken.

**Review by TomerShor — APPROVED** [review]
👌

File: pipeline-adapters/mlrun-pipelines-kfp-common/src/mlrun_pipelines/common/ops.py [hunk 3]
```diff
@@ -663,25 +663,13 @@ def _sanitize_ui_metadata(struct):
...
-            config_node_selector=config_node_selector,
-        )
-        return mlrun.utils.helpers.to_non_empty_values_dict(
-            mlrun.utils.helpers.merge_dicts_with_precedence(
-                config_node_selector,
-                project.spec.default_function_node_selector,
-                function_node_selector,
-            )
-        )
+    function_node_selector = mlrun.utils.helpers.enrich_node_selectors(
```
Comment (liranbg) [inline pipeline-adapters/mlrun-pipelines-kfp-common/src/mlrun_pipelines/common/ops.py]: Don't forget to bump version on setup py

**Review by liranbg — CHANGES_REQUESTED** [review]

File: mlrun/utils/helpers.py [hunk 4]
```diff
@@ -1687,6 +1687,24 @@ def validate_component_version_compatibility(
     return True
 
 
+def enrich_node_selectors(
```
Comment (alonmr) [inline mlrun/utils/helpers.py]: move to `mlrun.runtimes.utils`

File: mlrun/utils/helpers.py [hunk 5]
```diff
@@ -1687,6 +1687,24 @@ def validate_component_version_compatibility(
...
+    project_node_selector: dict, runtime_node_selector: dict
```
Comment (alonmr) [inline mlrun/utils/helpers.py]: here it says `runtime_node_selector` but the log says `Enriching run`.
it should be generic so you don't need to specify it its a runtime or a run

File: server/api/crud/runtimes/nuclio/function.py [hunk 6]
```diff
@@ -493,6 +493,20 @@ def _resolve_and_set_build_requirements_and_commands(function, config):
     mlrun.utils.update_in(config, "spec.build.commands", commands)
 
 
+def _enrich_function_node_selector_with_project(
```
Comment (alonmr) [inline server/api/crud/runtimes/nuclio/function.py]: you are also enriching with service level here
```suggestion
def _enrich_function_node_selector(
```

**Review by alonmr — COMMENTED** [review]
Great job! Minor naming comments

**Review by rokatyy — APPROVED** [review]

**Review by alonmr — APPROVED** [review]
🔥

File: server/api/crud/runtimes/nuclio/function.py [hunk 7]
```diff
@@ -493,6 +493,18 @@ def _resolve_and_set_build_requirements_and_commands(function, config):
     mlrun.utils.update_in(config, "spec.build.commands", commands)
 
 
+def _resolve_node_selector(run_db, project_name, function_node_selector):
+    project_node_selector = {}
+
+    if run_db and project_name:
```
Comment (liranbg) [inline server/api/crud/runtimes/nuclio/function.py:499]: when will run_db be none? mind commenting the flow?

**Review by liranbg — COMMENTED** [review]
really minor

**Review by liranbg — APPROVED** [review]

**Pull request merged**